src/Cassandra/Requests/GraphRequestHandler.cs

```diff
@@ -0,0 +1,226 @@
+            var row = rowSet.FirstOrDefault();
+            if (row == null)
+            {
+                GraphRequestHandler.Logger.Verbose(
+                    "Empty response querying graph analytics server, query will not be routed optimally");
+                return statement;
+            }
+            var resultField = row.GetValue<IDictionary<string, string>>("result");
+            if (resultField == null || !resultField.ContainsKey("location") || resultField["location"] == null)
+            {
+                GraphRequestHandler.Logger.Verbose(
+                    "Could not extract graph analytics server location from RPC, query will not be routed optimally");
+                return statement;
+            }
+            var location = resultField["location"];
+            var hostName = location.Substring(0, location.LastIndexOf(':'));
+            var address = _session.Cluster.Configuration.AddressTranslator.Translate(
+                new IPEndPoint(IPAddress.Parse(hostName), _session.Cluster.Configuration.ProtocolOptions.Port));
+            var host = _session.Cluster.GetHost(address);
+            statement.PreferredHost = host;
+            return statement;
+        }
+
+        private GraphResultSet CreateGraphResultSet(RowSet rs, IGraphTypeSerializer serializer)
+        {
+            return GraphResultSet.CreateNew(
+                rs,
+                serializer.GraphProtocol,
+                serializer.GetGraphRowParser());
+        }
+
+        private ConvertedStatementResult GetIStatement(IGraphStatement graphStmt, GraphOptions options)
+        {
+            var graphProtocol = graphStmt.GraphProtocolVersion ?? options.GraphProtocolVersion;
+
+            if (graphProtocol == null)
+            {
+                throw new DriverInternalError("Unable to determine graph protocol version. This is a bug, please report.");
+            }
+
+            // Existing graph statement implementations of this method are empty
+            // but it's part of the public interface definition
+            var stmt = graphStmt.ToIStatement(options);
+            if (stmt != null)
+            {
+                return new ConvertedStatementResult
+                {
+                    Serializer = _graphTypeSerializerFactory.CreateSerializer(null, null, graphProtocol.Value, true),
+                    Statement = stmt
+                };
+            }
+
+            return ConvertGraphStatement(graphStmt, options, graphProtocol.Value);
+        }
+
+        private ConvertedStatementResult ConvertGraphStatement(
+            IGraphStatement graphStmt, GraphOptions options, GraphProtocol graphProtocol)
+        {
+            string jsonParams;
+            string query;
+            IGraphTypeSerializer serializer;
+
+            if (graphStmt is SimpleGraphStatement simpleGraphStatement)
+            {
+                serializer = _graphTypeSerializerFactory.CreateSerializer(null, null, graphProtocol, true);
```

Comment:
The issue here is with the way we pass the fluent driver's serializers to the core driver, they are passed through the new `FluentGraphStatement` object so we don't have a way to pass these serializers once and store them in a singleton object. Currently one graph type serializer instance is created per request. 

I think I can do a couple of changes to avoid creating the `CustomGraphSONXReader` and `CustomGraphSONXWriter` instances and reusing the same ones but to reuse the actual `GraphTypeSerializer` instance we need to change the way serializers are shared between drivers. We could store them in a concurrent dictionary or bag but I'm not sure if that would be faster than just creating a new `GraphTypeSerializer` instance and storing those custom serializers as fields (the way it is currently).

Reply:
Another potential solution would be to add a way for the fluent driver to "inject" those shared serializers into the core driver's session and store which sessions have already been "injected" with these serializers but the fluent driver would have no way to know when a session is disposed so this collection could potentially grow to infinity.

Reply:
Or we could also just "assume" that the custom serializers are always the same and the graph type serializer would store these the first time a fluent graph statement is executed but it seems a bit too hacky maybe?

Reply:
Oh there's another option which is kind of hacky too but it might be the best one, we could have a static collection of these serializers that is populated by the fluent driver and each session's graph type serializer uses those static collections. They would be public but kind of hidden from the user, at least much more hidden than the `FluentGraphStatement` approach that receives them in the constructor (I really don't want to go with the `InternalsVisibleTo` approach 😢 )

`Cassandra.DataStax.Internal.GlobalGraphSerializers.Add(fluentDriverSerializers);`

Reply:
I think that, as long as the `FluentGraphStatement` instances use the same reference to the `customSerializers` and `customDeserializers` tuple, we can reuse it.

We could have a dictionary of custom serializers and deserializers tuple keys and IGraphTypeSerializer values.

Reply:
Done, added a dictionary with a custom `CacheKey` that takes into account every constructor argument of the `GraphTypeSerializer`

Reply:
Nice!